Lib/test/support/__init__.py

```diff
@@ -1900,21 +1899,54 @@ def _run_suite(suite):
         raise TestFailed(err)
 
 
-def _match_test(test):
-    global match_tests
+# By default, don't filter tests
+_match_test_func = None
 
-    if match_tests is None:
+
+def match_test(test):
+    # Function used by support.run_unittest() and regrtest --list-cases
+    if _match_test_func is None:
         return True
-    test_id = test.id()
+    else:
+        return _match_test_func(test.id())
+
+
+def _is_full_match_test(pattern):
+    # If a pattern contains at least one dot, it's considered as a full test
+    # identifier like 'test.test_os.FileTests.test_access'.
+    #
+    # Reject patterns which contain fnmatch patterns: '*', '?', '[...]'
+    # or '[!...]'. For example, reject 'test_access*'.
+    return ('.' in pattern) and all(char not in '?*[]' for char in pattern)
```

Comment:
I'm wondering whether using a regular expression would be faster.
```python
return ('.' in pattern) and not re.search(r'[?*\[\]]', pattern)
```

Reply:
I think that (char not in '?*[]') is more readable, and this function doesn't seem to be a performance bottleneck, there is no need to optimize it. Moreover, I just added a cache to not recompile patterns if set_match_tests() is called twice with the same patterns.

It seems like the compilation to a giant regex is slow, but I don't think that we can optimize this much, and again, I don't think that it's worth it. My main usage of many patterns is when I use test.bisect, but test.bisect only uses full test identifiers and so should use set().__contains__.